File: src/agents/maintenance/dependency-health.ts
import { execSync } from 'child_process';
import type { MaintenanceFinding } from '../shared/types';

type NpmOutdatedInfo = {
  current?: string;
  wanted?: string;
  latest?: string;
};

export async function checkDependencyHealth(): Promise<MaintenanceFinding[]> {
  const findings: MaintenanceFinding[] = [];
  const output = getNpmOutdatedOutput();

  if (!output) {
    return findings;
  }

  const outdated = JSON.parse(output || '{}') as Record<string, NpmOutdatedInfo>;

  for (const [pkg, info] of Object.entries(outdated)) {
    if (!info.current || !info.latest) {
      continue;
    }

    const isPlaywright = pkg.includes('playwright');

    findings.push({
      type: 'dependency_update',
      severity: isPlaywright ? 'high' : 'low',
      message: `${pkg} is outdated. Current: ${info.current}, latest: ${info.latest}`,
      payload: {
        package: pkg,
        current: info.current,
        wanted: info.wanted,
        latest: info.latest
      }
    });
  }

  return findings;
}

function getNpmOutdatedOutput(): string {
  try {
    return execSync('npm outdated --json', {
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'ignore']
    });
  } catch (error) {
    const npmError = error as { stdout?: string | Buffer };
    return npmError.stdout?.toString() ?? '';
  }
}

File: python_agents/agents/maintenance/dependency_health.py
from __future__ import annotations

import json
import os
import subprocess

from ..shared.models import Finding


def check_dependency_health() -> list[Finding]:
    output = _npm_outdated_output()
    if not output:
        return []

    try:
        outdated = json.loads(output)
    except json.JSONDecodeError as error:
        return [
            Finding(
                type="dependency_update",
                severity="low",
                message="Unable to parse npm outdated output",
                payload={"error": str(error)},
            )
        ]

    findings: list[Finding] = []
    for package_name, info in outdated.items():
        current = info.get("current")
        latest = info.get("latest")
        if not current or not latest:
            continue

        is_playwright_related = "playwright" in package_name.lower()
        findings.append(
            Finding(
                type="dependency_update",
                severity="high" if is_playwright_related else "low",
                message=f"{package_name} is outdated. Current: {current}, latest: {latest}",
                payload={
                    "package": package_name,
                    "current": current,
                    "wanted": info.get("wanted"),
                    "latest": latest,
                },
            )
        )

    return findings


def _npm_outdated_output() -> str:
    try:
        completed = subprocess.run(
            [_npm_command(), "outdated", "--json"],
            capture_output=True,
            text=True,
            timeout=90,
            check=False,
        )
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return ""

    return completed.stdout.strip()


def _npm_command() -> str:
    return "npm.cmd" if os.name == "nt" else "npm"
